beetsplug/vgmdb.py

```diff
@@ -0,0 +1,248 @@
+        dist = Distance()
+        if album_info.data_source == 'VGMdb':
+            dist.add('source', self.source_weight)
+        return dist
+
+    def candidates(self, items, artist, album, va_likely, extra_tags=None):
+        """Returns a list of AlbumInfo objects for VGMdb search results
+        matching an album and artist (if not various).
+        """
+        if va_likely:
+            query = album
+        else:
+            query = '%s %s' % (artist, album)
+        try:
+            ret = self.get_albums(query, va_likely)
+            for x in ret:
+                msg = 'Found album: %s ; id: %s ; artist: %s' % (x.album, x.album_id, x.artist)
+                log.debug(msg)
+            return ret
+        except:
+            log.debug('VGMdb Search Error: (query: %s)' % query)
+            return []
+
+    def album_for_id(self, album_id):
+        """Fetches an album by its VGMdb ID and returns an AlbumInfo object
+        or None if the album is not found.
+        """
+        log.debug('Querying VGMdb for release %s' % str(album_id))
+
+        # Get from VGMdb
+        r = requests.get('http://vgmdb.info/album/%s?format=json' % str(album_id))
+
+        # Decode Response's content
+        try:
+            item = r.json()
+        except:
+            log.debug('VGMdb JSON Decode Error: (id: %s)' % album_id)
+            return None
+
+        return self.get_album_info(item, False)
+
+    def get_albums(self, query, va_likely):
+        """Returns a list of AlbumInfo objects for a VGMdb search query.
+        """
+        # Strip non-word characters from query. Things like "!" and "-" can
+        # cause a query to return no results, even if they match the artist or
+        # album title. Use `re.UNICODE` flag to avoid stripping non-english
+        # word characters.
+        query = re.sub(r'(?u)\W+', ' ', query)
+        # Strip medium information from query, Things like "CD1" and "disk 1"
+        # can also negate an otherwise positive result.
+        query = re.sub(r'(?i)\b(CD|disc)\s*\d+', '', query)
+
+        # Query VGMdb
+        r = requests.get('http://vgmdb.info/search/albums/%s?format=json' % query)
+        albums = []
+
+        # Decode Response's content
+        try:
+            items = r.json()
+        except:
+            log.debug('VGMdb JSON Decode Error: (query: %s)' % query)
+            return albums
+
+        # Break up and get search results
+        for item in items["results"]["albums"]:
+            album_id = str(self.decod(item["link"][6:]))
+            albums.append(self.album_for_id(album_id))
+            if len(albums) >= 5:
+                break
+        log.debug('get_albums Querying VGMdb for release %s' % str(query))
+        return albums
+
+    def decod(self, val, codec='utf8'):
+        """Ensure that all string are coded to Unicode.
+        """
+        if isinstance(val, str):
+            b = val.encode()
+            return b.decode(codec, 'ignore')
```

Comment:
I'm a little unclear about the purpose here. If it's already a `str`, why bother encoding it and decoding again? In what cases is this invoked on something that is *not* a `str`, and why would it be left unchanged?